src/ray/util/event_test.cc

```diff
@@ -55,14 +55,110 @@ void CheckEventDetail(rpc::Event &event, std::string job_id, std::string node_id
+    EXPECT_EQ(event.label(), label);
+  }
+
+  if (message != "NULL") {
+    EXPECT_EQ(event.message(), message);
+  }
+
   EXPECT_EQ(event.source_pid(), getpid());
 }
 
+rpc::Event GetEventFromString(std::string seq) {
+  rpc::Event event;
+  std::vector<std::string> splitArray;
+  int pre_point = -1;
+  for (int i = 0, len = seq.length(); i < len; ++i) {
+    if (seq[i] == '\"') {
+      if (pre_point != -1) {
+        splitArray.push_back(seq.substr(pre_point + 1, i - 1 - pre_point));
+        pre_point = -1;
+      } else {
+        pre_point = i;
+      }
+    }
+  }
+  EXPECT_EQ(splitArray[0], "time_stamp");
+  event.set_timestamp(splitArray[1]);
+  EXPECT_EQ(splitArray[2], "severity");
+  rpc::Event_Severity severity_ele =
+      rpc::Event_Severity::Event_Severity_Event_Severity_INT_MIN_SENTINEL_DO_NOT_USE_;
+  RAY_CHECK(rpc::Event_Severity_Parse(splitArray[3], &severity_ele));
+  event.set_severity(severity_ele);
+  EXPECT_EQ(splitArray[4], "label");
+  event.set_label(splitArray[5]);
+  EXPECT_EQ(splitArray[6], "event_id");
+  event.set_event_id(splitArray[7]);
+  EXPECT_EQ(splitArray[8], "source_type");
+  rpc::Event_SourceType source_type_ele = rpc::Event_SourceType::
+      Event_SourceType_Event_SourceType_INT_MIN_SENTINEL_DO_NOT_USE_;
+  RAY_CHECK(rpc::Event_SourceType_Parse(splitArray[9], &source_type_ele));
+  event.set_source_type(source_type_ele);
+  EXPECT_EQ(splitArray[10], "host_name");
+  event.set_source_hostname(splitArray[11]);
+  EXPECT_EQ(splitArray[12], "pid");
+  event.set_source_pid(std::stoi(splitArray[13].c_str()));
+  EXPECT_EQ(splitArray[14], "message");
+  event.set_message(splitArray[15]);
+  EXPECT_EQ(splitArray[16], "custom_fields");
+  std::unordered_map<std::string, std::string> custom_fields;
+  for (int i = 17, len = splitArray.size(); i < len; i += 2) {
+    custom_fields[splitArray[i]] = splitArray[i + 1];
+  }
+  event.mutable_custom_fields()->insert(custom_fields.begin(), custom_fields.end());
+  return event;
+}
+
+void ParallelRunning(int nthreads, int loop_times,
+                     std::function<void()> event_context_init,
+                     std::function<void(int)> loop_function) {
+  if (nthreads > 1) {
+    std::vector<std::thread> threads(nthreads);
+    for (int t = 0; t < nthreads; t++) {
+      threads[t] = std::thread(std::bind(
+          [&](const int bi, const int ei, const int t) {
+            event_context_init();
+            for (int loop_i = bi; loop_i < ei; loop_i++) {
+              loop_function(loop_i);
+            }
+          },
+          t * loop_times / nthreads,
+          (t + 1) == nthreads ? loop_times : (t + 1) * loop_times / nthreads, t));
+    }
+    std::for_each(threads.begin(), threads.end(), [](std::thread &x) { x.join(); });
+  } else {
+    event_context_init();
+    for (int loop_i = 0; loop_i < loop_times; loop_i++) {
+      loop_function(loop_i);
+    }
+  }
+}
+
+std::vector<std::string> ReadEventFromFile(std::string log_file) {
```

Comment:
It's not suggested to return a right value vector.

Reply:
Do you mean use 
```C++
const std::vector<std::string>& vc = ReadEventFromFile(log_dir + "/event_RAYLET.log");
```
Or modify the function to 
```C++
ReadEventFromFile(vector<string>& vc, string filename)
```

Reply:
> Do you mean use
> 
> ```c++
> const std::vector<std::string>& vc = ReadEventFromFile(log_dir + "/event_RAYLET.log");
> ```
> 
> Or modify the function to
> 
> ```c++
> ReadEventFromFile(vector<string>& vc, string filename)
> ```

We prefer to second definition. But you need comment what input parameters and output parameters.